extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthServiceExtension.java

```diff
@@ -1,41 +1,76 @@
 package org.dataportabilityproject.auth.microsoft;
 
 import com.google.common.collect.ImmutableList;
+import java.io.IOException;
 import java.util.List;
+import okhttp3.OkHttpClient;
 import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
 import org.dataportabilityproject.spi.gateway.auth.AuthDataGenerator;
-import org.dataportabilityproject.spi.gateway.auth.extension.AuthServiceExtension;
 import org.dataportabilityproject.spi.gateway.auth.AuthServiceProviderRegistry.AuthMode;
+import org.dataportabilityproject.spi.gateway.auth.extension.AuthServiceExtension;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** */
 public class MicrosoftAuthServiceExtension implements AuthServiceExtension {
-  private MicrosoftAuthDataGenerator contactsGenerator;
+  private static final ImmutableList<String> DEFAULT_AUTH_GENERATOR_SERVICES =
+      ImmutableList.of("contacts", "calendar");
+  private static final Logger logger = LoggerFactory.getLogger(MicrosoftAuthServiceExtension.class);
 
-  public MicrosoftAuthServiceExtension() {
-    // contactsGenerator = new MicrosoftAuthDataGenerator(r)
-  }
+  private volatile MicrosoftAuthDataGenerator authDataGenerator;
+
+  public MicrosoftAuthServiceExtension() {}
 
   public String getServiceId() {
     return "microsoft";
   }
 
   public AuthDataGenerator getAuthDataGenerator(String transferDataType, AuthMode mode) {
-    // TODO
-    throw new UnsupportedOperationException();
+    // TODO Create auth data generator for a given mode, usually means a different scope
+    if (DEFAULT_AUTH_GENERATOR_SERVICES.contains(transferDataType)) {
+      return authDataGenerator;
+    }
+    throw new UnsupportedOperationException("transferDataType does not have an AuthDataGenerator");
   }
 
   @Override
   public List<String> getImportTypes() {
-    return ImmutableList.of();
+    return DEFAULT_AUTH_GENERATOR_SERVICES;
   }
 
   @Override
   public List<String> getExportTypes() {
-    return ImmutableList.of();
+    return DEFAULT_AUTH_GENERATOR_SERVICES;
   }
 
   @Override
   public void initialize(ExtensionContext context) {
-    // TODO
+    makeAuthDataGenerator(context);
+  }
+
+  private synchronized MicrosoftAuthDataGenerator makeAuthDataGenerator(ExtensionContext context) {
+    if (authDataGenerator == null) {
+      AppCredentials credentials;
+      try {
+        credentials =
+            context
+                .getService(AppCredentialStore.class)
+                .getAppCredentials("MICROSOFT_KEY", "MICROSOFT_SECRET");
```

Comment:
something to think about for another PR probably, in the modular structure maybe we could do some kind of enforcement that only Microsoft accesses the microsoft keys. This could prevent a malicious integrator from redirecting the user to, say, a Google oauth consent screen and tricking them into authorizing access to their Google data. The user would have to explicitly grant permission still, but as we saw with that Google docs oauth scam last year (https://www.darkreading.com/attacks-breaches/google-docs-phishing-scam-a-game-changer/d/d-id/1328808?) people can be tricked into these things pretty easily.